L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorAutoEncoderImpl.cc

```diff
@@ -0,0 +1,193 @@
+#include "L1Trigger/L1THGCal/interface/concentrator/HGCalConcentratorAutoEncoderImpl.h"
+#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
+#include "DataFormats/ForwardDetId/interface/HGCalTriggerDetId.h"
+#include <iomanip>
+
+// Following example of implementing graphloading from here:
+// https://gitlab.cern.ch/mrieger/CMSSW-TensorFlowExamples/-/blob/master/GraphLoading/
+
+HGCalConcentratorAutoEncoderImpl::HGCalConcentratorAutoEncoderImpl(const edm::ParameterSet& conf)
+    : cellRemap_(conf.getParameter<std::vector<int>>("cellRemap")),
+      bitsPerInput_(conf.getParameter<int>("nBitsPerInput")),
+      maxBitsPerOutput_(conf.getParameter<int>("maxBitsPerOutput")),
+      outputBitsPerLink_(conf.getParameter<std::vector<int>>("bitsPerLink")),
+      modelFilePaths_(conf.getParameter<std::vector<edm::ParameterSet>>("modelFiles")),
+      inputTensorName_encoder_(conf.getParameter<std::string>("encoderInputTensorName")),
+      outputTensorName_encoder_(conf.getParameter<std::string>("encoderOutputTensorName")),
+      inputTensorName_decoder_(conf.getParameter<std::string>("decoderInputTensorName")),
+      outputTensorName_decoder_(conf.getParameter<std::string>("decoderOutputTensorName")),
+      linkToGraphMap_(conf.getParameter<std::vector<int>>("linkToGraphMap")) {
+  //construct inverse array, to get U/V for a particular ae output position
+  for (unsigned i = 0; i < cellRemap_.size(); i++) {
+    if (cellRemap_.at(i) > -1) {
+      ae_outputCellU_[cellRemap_.at(i)] = int(i / 8);
+      ae_outputCellV_[cellRemap_.at(i)] = i % 8;
+    }
+  }
+
+  tensorflow::setLogging("0");
+
+  for (const auto& modelFilePset : modelFilePaths_) {
+    string encoderPath = modelFilePset.getParameter<edm::FileInPath>("encoderModelFile").fullPath();
+    string decoderPath = modelFilePset.getParameter<edm::FileInPath>("decoderModelFile").fullPath();
+
+    std::cout << "loading encoder graph from " << encoderPath << std::endl;
+    graphDef_encoder_ = tensorflow::loadGraphDef(encoderPath);
+
+    // create a new session and add the graphDef
+    session_encoder_.push_back(tensorflow::createSession(graphDef_encoder_));
+
+    std::cout << "loading decoder graph from " << decoderPath << std::endl;
+    graphDef_decoder_ = tensorflow::loadGraphDef(decoderPath);
+
+    // create a new session and add the graphDef
+    session_decoder_.push_back(tensorflow::createSession(graphDef_decoder_));
+  }
+}
+
+void HGCalConcentratorAutoEncoderImpl::select(unsigned nLinks,
```

Comment:
An exception should be thrown if the module size is >1 (which is still possible for decentralized configurations, although this is not used anymore).